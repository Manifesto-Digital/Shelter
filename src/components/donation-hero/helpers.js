export const formatDonationContent = content => {
  const createOptionObject = (amount, description, image) => ({
    amount,
    description,
    image,
  });

  // single
  const { singleAmount1, singleDescription1, singleImage1 } = content;
  const singleOption1 = createOptionObject(
    singleAmount1,
    singleDescription1,
    singleImage1
  );
  const { singleAmount2, singleDescription2, singleImage2 } = content;
  const singleOption2 = createOptionObject(
    singleAmount2,
    singleDescription2,
    singleImage2
  );
  const { singleAmount3, singleDescription3, singleImage3 } = content;
  const singleOption3 = createOptionObject(
    singleAmount3,
    singleDescription3,
    singleImage3
  );
  const { singleAmount4, singleDescription4, singleImage4 } = content;
  const singleOption4 = createOptionObject(
    singleAmount4,
    singleDescription4,
    singleImage4
  );

  const { singleAmount5, singleDescription5, singleImage5 } = content;
  const singleOption5 = createOptionObject(
    singleAmount5,
    singleDescription5,
    singleImage5
  );

  // monthly
  const { monthlyAmount1, monthlyDescription1, monthlyImage1 } = content;
  const monthlyOption1 = createOptionObject(
    monthlyAmount1,
    monthlyDescription1,
    monthlyImage1
  );
  const { monthlyAmount2, monthlyDescription2, monthlyImage2 } = content;
  const monthlyOption2 = createOptionObject(
    monthlyAmount2,
    monthlyDescription2,
    monthlyImage2
  );
  const { monthlyAmount3, monthlyDescription3, monthlyImage3 } = content;
  const monthlyOption3 = createOptionObject(
    monthlyAmount3,
    monthlyDescription3,
    monthlyImage3
  );
  const { monthlyAmount4, monthlyDescription4, monthlyImage4 } = content;
  const monthlyOption4 = createOptionObject(
    monthlyAmount4,
    monthlyDescription4,
    monthlyImage4
  );
  return {
    single: [
      singleOption1,
      singleOption2,
      singleOption3,
      singleOption4,
      singleOption5,
    ],
    monthly: [monthlyOption1, monthlyOption2, monthlyOption3, monthlyOption4],
  };
};
