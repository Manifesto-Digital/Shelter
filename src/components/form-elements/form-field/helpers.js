const possibleTextInputs = Object.freeze([
  'email',
  'text-field',
  'phone-number',
  'numeric',
  'text-area',
]);

export const isTextField = fieldType => possibleTextInputs.includes(fieldType);

export const getTextInputType = fieldType => {
  if (fieldType === 'phone-number') return 'tel';
  if (fieldType === 'numeric') return 'number';
  if (fieldType === 'email') return fieldType;
  if (fieldType === 'text-area') return 'textarea';
  return 'text';
};
