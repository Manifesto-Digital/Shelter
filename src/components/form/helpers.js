import * as Yup from 'yup';
import { consistentString } from '../../utils/content-formatting';

const isMultipleCheckbox = field =>
  consistentString(field.fieldType) === 'checkboxes' &&
  field.defaultValue &&
  field.valueOptions.length > 1;

export const getInitialValues = (fields, hiddenInitialValues) => {
  const initialValues = hiddenInitialValues;

  // Loop through all fields to build an object to store initial values to pass to formik
  fields.forEach(field => {
    if (!field.fieldType) return;

    if (field.fieldType === 'Address') return; // Temporary

    // Set default value to be the initial value for form
    if (field.internal.type === 'ContentfulTopicFormField') {
      initialValues[field.machineName] = field.defaultValue || '';
    }
    // Multiple checkboxes require an array of values
    if (isMultipleCheckbox(field)) {
      initialValues[field.machineName] = field.defaultValue
        ? [field.defaultValue]
        : [];
    }

    if (field.internal.type === 'ContentfulTopicFormFieldset') {
      getInitialValues(field.formFields);
    }
  });
  return initialValues;
};

export const getValidationSchema = formFields => {
  const validationSchema = {};

  // Loop through all form fields and build a Yup validationSchema object for formik
  formFields.forEach(field => {
    if (!field.fieldType) return;
    const fieldType = consistentString(field.fieldType);
    if (field.fieldType === 'address') return; // Temporary
    const fieldName = field.machineName;

    let validationType;
    // Apply a validation type to each field
    if (isMultipleCheckbox(field)) validationType = Yup.array();
    else if (fieldType === 'phone-number') {
      validationType = Yup.number().typeError('Please provide a valid number');
    } else if (fieldType === 'email') {
      validationType = Yup.string().email('Please provide a valid email');
    } else validationType = Yup.string();

    // Add required validation if specified
    if (field.required) validationType = validationType.required('Required');

    validationSchema[fieldName] = validationType;
  });

  return Yup.object().shape(validationSchema);
};
