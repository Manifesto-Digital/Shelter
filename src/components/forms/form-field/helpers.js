const possibleTextInputs = Object.freeze([
  'email',
  'text-field',
  'phone-number',
  'numeric',
]);

export const isTextField = fieldType => possibleTextInputs.includes(fieldType);

export const getTextInputType = fieldType => {
  if (fieldType === 'phone-number') return 'tel';
  if (fieldType === 'numeric') return 'number';
  if (fieldType === 'email') return fieldType;
  return 'text';
};
