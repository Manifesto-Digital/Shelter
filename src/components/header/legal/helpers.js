/**
 *
 * Loop through the professionals navigation menu to find the current active page
 * and the parent pages so to that the correct menus and active states will be shown
 *
 * @param {Object []} childMenuItems
 * @param {Object} location
 *
 * @returns [menuParentSlug, menuParentSlug, activePageSlug] (example)
 */
export const getActivePages = (childMenuItems, location) => {
  const isActivePage = slug => {
    // Match optional trailing slash
    const string = `${slug}((/w+)+|/?)$`;
    const regex = new RegExp(string, 'g');
    return location.pathname.match(regex);
  };

  let activePagesArray = [];
  let pageFound = false;

  const checkLevel = items => {
    for (let i = 0; i < items.length; i += 1) {
      const item = items[i];
      if (!item.navigationLink) return;

      const itemSlug = item.navigationLink[0].slug;

      if (isActivePage(itemSlug)) {
        // If found push the active page slug
        activePagesArray.push(itemSlug);
        pageFound = true;
        break;
      } else if (item.childNavigationItems) {
        // Recursively go through the child navigation items storing
        // potential active parents item slugs
        activePagesArray.push(itemSlug);
        checkLevel(item.childNavigationItems);
      }
    }
    if (!pageFound) activePagesArray = [];
    return activePagesArray;
  };

  checkLevel(childMenuItems);
  return activePagesArray;
};
