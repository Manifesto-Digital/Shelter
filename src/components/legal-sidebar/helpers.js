export const buildCurrentPageHierarchy = (hierarchy, parentSlug) => {
  const hierarchyDepth = 2;

  // Get the parent of the topmost level to show for use in the heading.
  const headingRef =
    hierarchyDepth > parentSlug.length
      ? parentSlug[parentSlug.length - hierarchyDepth - 1]
      : parentSlug[0];
  const heading = headingRef.label;

  // The top most level is Legal which we do not want to show so remove from parentSlug
  const legalChildrenSlugs = parentSlug.slice(1);

  // Slice the pages parent slug if the current page greater than hierarchy depth
  const topLevelToShowIndex = legalChildrenSlugs.length - hierarchyDepth;
  const parentHierarchyToShow =
    legalChildrenSlugs.length > hierarchyDepth
      ? legalChildrenSlugs.slice(topLevelToShowIndex)
      : legalChildrenSlugs;

  // Check if page exists on current hierarchy level
  function existsOnLevel(pageRef) {
    return parentHierarchyToShow.find(ref => ref.slug === pageRef.slug);
  }

  // Copy to avoid mutating original hierarchy
  const hierarchyCopy = JSON.parse(JSON.stringify(hierarchy));

  // Loop through assign an active state to the current page parents
  function buildActiveHierarchy(hierarchyLevel) {
    const newHierarchy = [];
    function recursiveLoop(level) {
      level.forEach((pageRef, i) => {
        if (existsOnLevel(pageRef)) {
          // Set current item to active
          level[i].active = true;
          // Push the first time this matches, this will also include the children
          if (newHierarchy.length === 0) newHierarchy.push(level);
        }
        if (pageRef.children) {
          recursiveLoop(pageRef.children);
        }
      });
    }
    recursiveLoop(hierarchyLevel);

    return newHierarchy;
  }

  const currentPageHierarchy = buildActiveHierarchy(hierarchyCopy);
  return { currentPageHierarchy: currentPageHierarchy[0], heading };
};
