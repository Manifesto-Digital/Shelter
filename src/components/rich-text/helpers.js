export const fieldsMap = fieldsToMap =>
  Object.keys(fieldsToMap).reduce((accum, key) => {
    const field = fieldsToMap[key];
    if (Object.prototype.hasOwnProperty.call(field, 'en-GB')) {
      accum[key] = flattenKey('en-GB', field); // eslint-disable-line no-use-before-define
    } else if (typeof field === 'object' && field !== null) {
      accum[key] = fieldsMap(field);
    } else {
      accum[key] = field;
    }
    return accum;
  }, {});

const flattenKey = (key, field) => {
  if (typeof field[key] === 'string' || typeof field[key] === 'boolean') {
    return field[key];
  }
  if (Array.isArray(field[key]) && field[key].length === 1) {
    return [fieldsMap(field[key][0].fields)];
  }
  if (typeof field[key] === 'object') {
    if (Object.prototype.hasOwnProperty.call(field[key], 'fields')) {
      return fieldsMap(field[key].fields);
    }

    return field[key];
  }

  return field;
};
