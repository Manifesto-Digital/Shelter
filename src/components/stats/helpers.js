export const statsToArray = content => {
  const createStatObject = (title, subtitle, text) => {
    if (!text) return null;

    return {
      title,
      subtitle,
      text,
    };
  };

  const { stat1Title, stat1Subtitle, stat1Text } = content;
  const stat1 = createStatObject(stat1Title, stat1Subtitle, stat1Text);
  const { stat2Title, stat2Subtitle, stat2Text } = content;
  const stat2 = createStatObject(stat2Title, stat2Subtitle, stat2Text);
  const { stat3Title, stat3Subtitle, stat3Text } = content;
  const stat3 = createStatObject(stat3Title, stat3Subtitle, stat3Text);
  const { stat4Title, stat4Subtitle, stat4Text } = content;
  const stat4 = createStatObject(stat4Title, stat4Subtitle, stat4Text);

  // Remove any nulls as a failsafe
  return [stat1, stat2, stat3, stat4].filter(stat => stat);
};
