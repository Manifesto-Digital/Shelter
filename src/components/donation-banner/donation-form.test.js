import React from 'react';
import { snapshotComponent, mountWithTheme } from '../../../__tests__/helpers';
import DonationForm from './donation-form';
import Button from '../button';

it('renders correctly', () => {
  snapshotComponent(<DonationForm />);
});

it('has all the necessary fields', () => {
  const fieldNames = [
    'cid',
    'free_amount',
    'amount',
    'reserved_appeal_code',
    'frequency',
    'amount-holder',
  ];
  const wrapper = mountWithTheme(<DonationForm />);
  fieldNames.forEach(name => {
    expect(wrapper.find(`input[name="${name}"]`)).toHaveLength(1);
  });
});

it('has a default value set in amount field', () => {
  const wrapper = mountWithTheme(<DonationForm />);
  const hiddenInput = wrapper.find('input[name="amount"]');
  expect(parseInt(hiddenInput.props().value)).toBeGreaterThan(0);
});

it('updates the hidden amount field when visible donation amount field changes', () => {
  const wrapper = mountWithTheme(<DonationForm />);
  const visibleInput = wrapper.find('input[name="amount-holder"]');

  visibleInput.simulate('change', { target: { value: '25' } });

  const hiddenInput = wrapper.find('input[name="amount"]');
  // Expect the amount hidden field to be the pence value of the visible field
  expect(hiddenInput.props().value).toEqual((25 * 100).toString());
});

it('reverts to default value when visible input has no value', () => {
  const mockDefaultValue = 20;
  const wrapper = mountWithTheme(
    <DonationForm defaultDonationValue={mockDefaultValue} />
  );
  const visibleInput = wrapper.find('input[name="amount-holder"]');
  // Give visible input value then remove
  visibleInput.simulate('change', { target: { value: '25' } });
  visibleInput.simulate('change', { target: { value: '' } });

  const hiddenInput = wrapper.find('input[name="amount"]');
  // Expect the amount hidden field to have reverted back to the default
  expect(hiddenInput.props().value).toEqual(
    (mockDefaultValue * 100).toString()
  );
});

it('has the correct form method', () => {
  const wrapper = mountWithTheme(<DonationForm />);
  // Check method is GET
  expect(wrapper.find('form').props().method).toEqual('GET');
});

it('has the correct form action', () => {
  const wrapper = mountWithTheme(<DonationForm />);
  // Check the action URL is correct
  expect(wrapper.find('form').props().action).toEqual(
    'https://donate.shelter.org.uk/b'
  );
});

it('has submit button', () => {
  const wrapper = mountWithTheme(<DonationForm />);
  expect(wrapper.find('button[type="submit"]')).toHaveLength(1);
});

it('shows the placeholder based on prop', () => {
  const mockDefaultValue = 30;
  const wrapper = mountWithTheme(
    <DonationForm defaultDonationValue={mockDefaultValue} />
  );
  expect(
    wrapper.find('input[name="amount-holder"]').props().placeholder
  ).toEqual(mockDefaultValue.toString());
});

it('sets button colour based on bg prop', () => {
  const wrapper = mountWithTheme(<DonationForm buttonBg="donate" />);

  expect(wrapper.find(Button).prop('bg')).toBe('donate');
});
