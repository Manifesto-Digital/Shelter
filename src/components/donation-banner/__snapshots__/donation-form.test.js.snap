// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c2 {
  position: relative;
}

.c4 {
  display: inline-block;
  background-size: 15px 15px;
  width: auto;
  margin-bottom: 0;
  padding: 10px;
  padding-left: 30px;
  max-width: 10em;
}

.c4::-webkit-outer-spin-button,
.c4::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c3 {
  left: 12px;
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  z-index: 1;
  font-size: 0.875rem;
}

.c1 {
  padding: 0;
  margin: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(1px);
  clip-path: inset(1px);
  width: 1px;
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
}

.c6 {
  display: inline-block;
  padding: 10px 1em;
  font-weight: normal;
  background-color: transparent;
  border: 1px solid #fff;
  color: #fff;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  margin-left: 10px;
}

@media (min-width:40.0625em) {
  .c0 {
    -webkit-flex: 0 1 40%;
    -ms-flex: 0 1 40%;
    flex: 0 1 40%;
    margin-bottom: 0;
    -webkit-box-pack: end;
    -webkit-justify-content: flex-end;
    -ms-flex-pack: end;
    justify-content: flex-end;
  }
}

<form
  action="https://donate.shelter.org.uk/b"
  className="c0"
  method="GET"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <input
    name="cid"
    onBlur={[Function]}
    onChange={[Function]}
    type="hidden"
    value="263"
  />
  <input
    name="free_amount"
    onBlur={[Function]}
    onChange={[Function]}
    type="hidden"
    value="1"
  />
  <input
    name="amount"
    onBlur={[Function]}
    onChange={[Function]}
    type="hidden"
    value="3000"
  />
  <input
    name="reserved_appeal_code"
    onBlur={[Function]}
    onChange={[Function]}
    type="hidden"
    value=""
  />
  <input
    name="frequency"
    onBlur={[Function]}
    onChange={[Function]}
    type="hidden"
    value="once"
  />
  <label
    className="c1"
    htmlFor="amount-holder"
  >
    Donate
  </label>
  <div
    className="c2"
  >
    <span
      className="c3"
    >
      Â£
    </span>
    <input
      className="c4"
      id="amount-holder"
      name="amount-holder"
      onBlur={[Function]}
      onChange={[Function]}
      pattern="[0-9]*"
      placeholder="30"
      type="number"
      value=""
    />
  </div>
  <button
    className="c5 c6"
    type="submit"
  >
    Donate
  </button>
</form>
`;
